staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -208,54 +400,46 @@ func (q *delayingType) waitingLoop() {
 		now := q.clock.Now()
 
 		// Add ready entries
+		popped := false
 		for waitingForQueue.Len() > 0 {
 			entry := waitingForQueue.Peek().(*waitFor)
 			if entry.readyAt.After(now) {
 				break
 			}
-
+			q.log(fmt.Sprintf("waitingLoop: item matured from waiting: %s", entry))
```

Comment:
The argument gets evaluated when you do it like this, which can be expensive.

Reply:
Hi, I've changed the log method to receive a func()string parameter instead of a string.
And then I've changed all of the log calls to pass a function that creates the log line instead of the actual log line.
It looks as though the function will only be called if q.logger is not nil, i.e. the expensive fmt.Sprintf etc will only be called when logging is enabled - which should give us good performance in general use and logging when we would benefit from it (in tests).

log() is simple enough to be inlined by the go compiler:
```
// ./delaying_queue.go:377:6: can inline (*delayingType).log with cost 66 as: method(*delayingType) func(func() string) { if q.logger != nil { (*delayingType).log_inner(q, lf) } }
```

Reply:
What do you think?  Would it be better to simply wrap all of the log lines inside a if q.logger != nil ?